src/librustc_resolve/resolve_imports.rs

```diff
@@ -1309,21 +1309,16 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
         if !is_redundant.is_empty() &&
             is_redundant.present_items().all(|is_redundant| is_redundant)
         {
-            self.session.buffer_lint(
+            self.session.buffer_lint_with_diagnostic(
                 REDUNDANT_IMPORT,
                 directive.id,
                 directive.span,
                 &format!("the item `{}` is imported redundantly", ident),
```

Comment:
Could you use "imported" only if `other_binding.is_import()` and "defined" otherwise, like it's done in other places where similar errors are reported.